File: app/src/main/java/com/karntrehan/posts/list/ListAdapter.java
package com.karntrehan.posts.list;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.karntrehan.posts.R;
import com.karntrehan.posts.databinding.PostItemBinding;
import com.karntrehan.posts.list.entity.Post;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by karn on 03-06-2017.
 */

public class ListAdapter extends RecyclerView.Adapter<ListAdapter.ViewHolder> {

    private ListContract.Presenter presenter;
    private List<Post> posts;

    public ListAdapter(ListContract.Presenter presenter) {
        this.presenter = presenter;
        posts = new ArrayList<>();
    }

    protected void setPosts(List<Post> posts) {
        this.posts = posts;
        notifyDataSetChanged();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        PostItemBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext())
                , R.layout.post_item, parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) {
        final Post post = posts.get(position);
        holder.binding.tvTitle.setText(post.getPostTitle());
        holder.binding.tvBody.setText(post.getFormattedPostBody());

        //SharedItem transition
        ViewCompat.setTransitionName(holder.binding.tvTitle, post.getPostTitle());
        ViewCompat.setTransitionName(holder.binding.tvBody, post.getFormattedPostBody());

        holder.binding.cvPost.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                presenter.postClicked(position,post,holder);
            }
        });
    }

    @Override
    public int getItemCount() {
        return posts.size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        PostItemBinding binding;

        public ViewHolder(PostItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: posts/src/main/java/com/karntrehan/posts/list/ListAdapter.kt
package com.karntrehan.posts.list

import android.support.v4.view.ViewCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.karntrehan.posts.R
import com.karntrehan.posts.commons.data.PostWithUser
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.post_item.view.*

class ListAdapter(private val picasso: Picasso) : RecyclerView.Adapter<ListAdapter.ListViewHolder>() {
    private var data = emptyList<PostWithUser>()
    var interactor: PostInteractor? = null

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): ListViewHolder {
        return ListViewHolder(LayoutInflater.from(parent?.context).inflate(R.layout.post_item, parent, false))
    }

    override fun onBindViewHolder(holder: ListViewHolder?, position: Int) {
        holder?.bind(data[position], picasso)

        if (holder?.itemView != null)
            holder.itemView.setOnClickListener { view ->
                with(view) {
                    interactor?.postClicked(data[position], tvTitle, tvBody, tvAuthorName, ivAvatar)
                }
            }
    }

    override fun getItemCount(): Int {
        return data.size
    }

    fun setData(data: List<PostWithUser>?) {
        this.data = data ?: this.data
        notifyDataSetChanged()
    }

    class ListViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        fun bind(post: PostWithUser, picasso: Picasso) {
            with(post) {
                itemView.tvTitle.text = postTitle
                itemView.tvBody.text = getFormattedPostBody()
                itemView.tvAuthorName.text = userName
                picasso.load(getAvatarPhoto()).into(itemView.ivAvatar)

                //SharedItem transition
                ViewCompat.setTransitionName(itemView.tvTitle, postTitle);
                ViewCompat.setTransitionName(itemView.tvBody, postBody);
                ViewCompat.setTransitionName(itemView.tvAuthorName, userName);
                ViewCompat.setTransitionName(itemView.ivAvatar, getAvatarPhoto());
            }
        }
    }

    interface PostInteractor {
        fun postClicked(post: PostWithUser, tvTitle: TextView, tvBody: TextView, tvAuthorName: TextView, ivAvatar: ImageView)
    }
}